### PR TITLE
Implement ClipboardItem.supports

### PR DESCRIPTION
**Pull request by karlcow**
#### 91237d0831bb66fe1c980301b4cbf62b72b51124
<pre>
Implement ClipboardItem.supports
<a href="https://bugs.webkit.org/show_bug.cgi?id=279712">https://bugs.webkit.org/show_bug.cgi?id=279712</a>
<a href="https://rdar.apple.com/136008522">rdar://136008522</a>

Reviewed by Wenson Hsieh.

Adds ClipboardItem.supports()
Adds also text/uri-lists per spec to the list of types in WPT
Defers custom web format implementation to &lt;<a href="https://webkit.org/b/280664">https://webkit.org/b/280664</a>&gt;
Defers support for image/svg+xml to &lt;<a href="https://webkit.org/b/280726">https://webkit.org/b/280726</a>&gt;

* LayoutTests/editing/async-clipboard/clipboard-interfaces-expected.txt: Removed.
* LayoutTests/editing/async-clipboard/clipboard-interfaces.html: Removed.
  These are covered by WPT

* LayoutTests/editing/async-clipboard/clipboard-write-basic-expected.txt:
* LayoutTests/editing/async-clipboard/clipboard-write-basic.html:
  Removed the invalid case of an empty clipboardItem()

* LayoutTests/imported/w3c/web-platform-tests/clipboard-apis/clipboard-item.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/clipboard-apis/clipboard-item.https.html:
  Fixing the tests. Needs to be exported back to WPT

* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/Modules/async-clipboard/ClipboardItem.cpp:
(WebCore::ClipboardItem::create):
(WebCore::ClipboardItem::supports):
* Source/WebCore/Modules/async-clipboard/ClipboardItem.h:
* Source/WebCore/Modules/async-clipboard/ClipboardItem.idl:

Canonical link: <a href="https://commits.webkit.org/284593@main">https://commits.webkit.org/284593@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/b6f364ba8d14710cb896d783d58e82493c2b79b1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69949 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49350 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22702 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21107 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20958 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/74034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13988 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73015 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60330 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/74034 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41622 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/17758 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19484 "Built successfully") | 
| | [loading 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63545 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; run-api-tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18109 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75749 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14174 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17335 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63204 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14210 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60405 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63140 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15520 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11160 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/4767 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45153 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46227 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47498 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45968 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/Modules/async-clipboard/ClipboardItem.cpp [hunk 1]
```diff
@@ -92,6 +98,20 @@ void ClipboardItem::getType(const String& type, Ref<DeferredPromise>&& promise)
...
 
+bool ClipboardItem::supports(const String& type)
+{
+    // FIXME: Custom format starts with `"web "`("web" followed by U+0020 SPACE) prefix
+    // and suffix (after stripping out `"web "`) passes the parsing a MIME type check.
+    if (type == textPlainContentTypeAtom()
+        || type == textHTMLContentTypeAtom()
+        || type == "image/png"_s
+        || type == "text/uri-list"_s
+        || type == "image/svg+xml"_s) {
```
Comment (whsieh) [inline Source/WebCore/Modules/async-clipboard/ClipboardItem.cpp]: Is this right? (since we don't support sanitized copy/paste of SVG data)

Comment (karlcow) [reply Source/WebCore/Modules/async-clipboard/ClipboardItem.cpp]: Hmm indeed. I need to remove and create a FIXME in there, probably with a bug on webkit.org.

Comment (karlcow) [reply Source/WebCore/Modules/async-clipboard/ClipboardItem.cpp]: Opened https://bugs.webkit.org/show_bug.cgi?id=280726

**Comment by karlcow** [comment]
@whsieh @rniwa  I wonder if `LayoutTests/editing/async-clipboard/` is still needed.
What are your thoughts?

File: LayoutTests/editing/async-clipboard/clipboard-interfaces.html [hunk 2]
```diff
@@ -11,7 +11,7 @@
             try {
                 testFailed(`${scriptToEvaluate} resolved to ${await eval(scriptToEvaluate)}`);
             } catch (exception) {
-                testPassed(`${scriptToEvaluate} rejected with ${exception}`);
+                testPassed(`${scriptToEvaluate} rejected with "${exception}"`);
```
Comment (karlcow) [inline LayoutTests/editing/async-clipboard/clipboard-interfaces.html]: copy pasta mistake. But waiting on if we should just delete these tests.

**Comment by karlcow** [comment]
hmm too many files in that PR.

**Comment by whsieh** [comment]
> @whsieh @rniwa I wonder if `LayoutTests/editing/async-clipboard/` is still needed. What are your thoughts?

Good point — `editing/async-clipboard/clipboard-interfaces.html` does seem redundant with the WPT, but the remaining tests should be kept because they exercise functionality that isn't expressible in WPT — for instance, iOS-specific keyboard interactions, paste menu UI.

**Review by whsieh — APPROVED** [review]
r=mews

**Comment by karlcow** [comment]
> > @whsieh @rniwa I wonder if `LayoutTests/editing/async-clipboard/` is still needed. What are your thoughts?
> 
> Good point — `editing/async-clipboard/clipboard-interfaces.html` does seem redundant with the WPT, but the remaining tests should be kept because they exercise functionality that isn't expressible in WPT — for instance, iOS-specific keyboard interactions, paste menu UI.

ok I removed the file covered by WPT, and I fixed one test in write-basic (because it is invalid)

**Comment by karlcow** [comment]
@whsieh only one failure for mac-wk1
https://ews-build.webkit.org/#/builders/64/builds/17646 

should I add an exception for this one?

**Comment by whsieh** [comment]
> @whsieh only one failure for mac-wk1 https://ews-build.webkit.org/#/builders/64/builds/17646
> 
> should I add an exception for this one?

Yeah, it seems fine to add a WK1-specific test expectation (baseline) for this.

We don't support the Clipboard API on WK1.

**Pull request merged**